File: algorithm_datastructure/leetcode/435.non-overlapping-intervals.py
#
# @lc app=leetcode id=435 lang=python3
#
# [435] Non-overlapping Intervals
#

# @lc code=start
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        size = len(intervals)
        if size <= 1:
            return 0
        # Sort the end intervals
        intervals.sort(key=lambda x: x[1])

        prev_end = intervals[0][1]
        ret = 0
        for i in range(1, size):
            start, end = intervals[i]
            if start >= prev_end:
                # No overlapping, update the current right boundary
                prev_end = end
            else:
                # There's overlapping
                ret += 1
        return ret


# @lc code=end

File: algorithm_data_structure/leetcode/435.non-overlapping-intervals.cpp
/*
 * @lc app=leetcode id=435 lang=cpp
 *
 * [435] Non-overlapping Intervals
 */

// @lc code=start

#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    int eraseOverlapIntervals(vector<vector<int>> &intervals)
    {
        // Check base case
        if (intervals.empty())
        {
            return 0;
        }
        // Sort the intevals with its end value. Use a lambda.
        sort(intervals.begin(), intervals.end(),
             [](vector<int> &a, vector<int> &b)
             {
                 return a[1] < b[1];
             });

        int removed = 0;
        int prev = intervals[0][1];

        // Start the interation from the second element
        for (int i = 1; i < intervals.size(); ++i)
        {
            if (intervals[i][0] < prev)
            {
                // Find a duplicated interval. The start value of the interval
                // is smaller than the end value of the previous interval
                ++removed;
            }
            else
            {
                // A unduplicated interval
                prev = intervals[i][1];
            }
        }

        return removed;
    }
};
// @lc code=end
